scispacy/candidate_generation.py

```diff
@@ -0,0 +1,307 @@
+        original_neighbours = self.ann_index.knnQueryBatch(vectors, k=k)
+
+        neighbors, distances = zip(*[(x[0].tolist(), x[1].tolist()) for x in original_neighbours])
+        neighbors = list(neighbors)
+        distances = list(distances)
+
+        # neighbors need to be convected to an np.array of objects instead of ndarray of dimensions len(vectors)xk
+        # Solution: add a row to `neighbors` with any length other than k. This way, calling np.array(neighbors)
+        # returns an np.array of objects
+        neighbors.append([])
+        distances.append([])
+        # interleave `neighbors` and Nones in `extended_neighbors`
+        extended_neighbors[empty_vectors_boolean_flags] = numpy.array(neighbors)[:-1]
+        extended_distances[empty_vectors_boolean_flags] = numpy.array(distances)[:-1]
+
+        return extended_neighbors, extended_distances
+
+    def generate_candidates(self, mention_texts: List[str], k: int) -> List[List[MentionCandidate]]:
+        """
+        Given a list of mention texts, returns a list of candidate neighbors.
+
+        NOTE: Because we include canonical name aliases in the ann index, the list
+        of candidates returned will not necessarily be of length k for each candidate,
+        because we then map these to canonical ids only.
+
+        NOTE: For a given mention, the returned candidate list might be empty, which implies that
+        the tfidf vector for this mention was all zeros (i.e there were no 3 gram overlaps). This
+        happens reasonably rarely, but does occasionally.
+        Parameters
+        ----------
+        mention_texts: List[str], required.
+            The list of mention strings to generate candidates for.
+        k: int, required.
+            The number of ann neighbours to look up.
+            Note that the number returned may differ due to aliases.
+
+        Returns
+        -------
+        A list of MentionCandidate objects per mention containing UMLS concept_ids and aliases
+        and distances which were mapped to. Note that these are lists for each concept id,
+        because the index contains aliases which are canonicalized, so multiple values may map
+        to the same canonical id.
+        """
+        if self.verbose:
+            print(f'Generating candidates for {len(mention_texts)} mentions')
+        tfidfs = self.vectorizer.transform(mention_texts)
+        start_time = datetime.datetime.now()
+
+        # `ann_index.knnQueryBatch` crashes if one of the vectors is all zeros.
+        # `nmslib_knn_with_zero_vectors` is a wrapper around `ann_index.knnQueryBatch` that addresses this issue.
+        batch_neighbors, batch_distances = self.nmslib_knn_with_zero_vectors(tfidfs, k)
+        end_time = datetime.datetime.now()
+        total_time = end_time - start_time
+        if self.verbose:
+            print(f'Finding neighbors took {total_time.total_seconds()} seconds')
+        neighbors_by_concept_ids = []
```

Comment:
naming of this variable suggests it is a dictionary with concept ids as the key. Could you add a comment clarifying or change the name?